Parsing issues on closure parameters and return type

This following line in libcore 1.49 throws our parser quite the curveball:
```rust
 let is_zero = &|&&d: &&u8| -> bool { d == b'0' };
 ```
 
 output:
 
 ```rust
 ../../rust/library/core/src/num/dec2flt/mod.rs:297:26: error: unrecognised token ‘&&’ in type
  297 |     let is_zero = &|&&d: &&u8| -> bool { d == b'0' };
      |                          ^~
../../rust/library/core/src/num/dec2flt/mod.rs:297:26: error: failed to parse type in closure parameter
../../rust/library/core/src/num/dec2flt/mod.rs:297:21: error: could not parse closure param
  297 |     let is_zero = &|&&d: &&u8| -> bool { d == b'0' };
      |                     ^~
../../rust/library/core/src/num/dec2flt/mod.rs:297:32: error: found unexpected token ‘->’ in null denotation
  297 |     let is_zero = &|&&d: &&u8| -> bool { d == b'0' };
      |                                ^~
../../rust/library/core/src/num/dec2flt/mod.rs:297:35: error: failed to parse expression in let statement
  297 |     let is_zero = &|&&d: &&u8| -> bool { d == b'0' };
      |                                   ^~~~
../../rust/library/core/src/num/dec2flt/mod.rs:297:5: error: failed to parse statement or expression without block in block expression
  297 |     let is_zero = &|&&d: &&u8| -> bool { d == b'0' };
      |     ^~~
  ```

--- a/gcc/rust/parse/rust-parse-impl.h
+++ b/gcc/rust/parse/rust-parse-impl.h
@@ -10949,7 +10949,7 @@ Parser<ManagedTokenSource>::parse_reference_pattern ()
     }
 
   // parse pattern to get reference of (required)
-  std::unique_ptr<AST::Pattern> pattern = parse_pattern ();
+  std::unique_ptr<AST::Pattern> pattern = parse_pattern_no_alt ();
   if (pattern == nullptr)
     {
       Error error (lexer.peek_token ()->get_locus (),

--- a/gcc/testsuite/rust/compile/issue-1807.rs
+++ b/gcc/testsuite/rust/compile/issue-1807.rs
@@ -1,0 +1,6 @@
+// { dg-additional-options "-fsyntax-only" }
+
+fn main() {
+    let is_zero = &|&&d: &&u8| -> bool { d == b'0' };
+    let lambda = |&c| c != b'9';
+}

